src/game/emutest.c

```diff
@@ -0,0 +1,168 @@
+extern u8 __osContLastCmd;
+extern void __osSiGetAccess(void);
+extern void __osSiRelAccess(void);
+
+enum Emulator gEmulator = EMU_CONSOLE;
+
+__attribute__((aligned(8)))
+static const u32 check_count_factor_asm[] = {
+    0x40084800u, // MFC0 T0, COUNT
+    0x40094800u, // MFC0 T1, COUNT
+    0x03E00008u, // JR RA
+    0x01281023u  // SUBU V0, T1, T0
+};
+
+static inline u32 check_count_factor() {
+    const u32 saved = __osDisableInt();
+    const u32 cf = ((u32 (*)(void))check_count_factor_asm)();
+    __osRestoreInt(saved);
+    return cf;
+}
+
+static inline enum Emulator get_pj64_version() {
+    // When calling this function, we know that the emulator is some version of Project 64,
+    // and it isn't using the PJ64 4.0 interpreter core. Figure out which version it is.
+    
+    // PJ64 4.0 dynarec core doesn't update the COUNT register correctly within recompiled functions
+    if (check_count_factor() == 0) {
+        return EMU_PROJECT64_4;
+    }
+
+    // Instead of implementing this PIF command correctly, PJ64 versions prior to 3.0 just have
+    // a set of hardcoded values for some requests. At least one of these hardcoded values has a
+    // typo in it, making it give an incorrect result.
+    __osSiGetAccess();
+    u32 *pifRam32 = (u32*)__osContPifRam;
+    for (s32 i = 0; i < 15; i++) pifRam32[i] = 0;
+    pifRam32[15] = 2;
+
+    const u8 cicTest[] = {
+        0x0F, 0x0F,
+        0xEC, 0x3C, 0xB6, 0x76, 0xB8, 0x1D, 0xBB, 0x8F,
+        0x6B, 0x3A, 0x80, 0xEC, 0xED, 0xEA, 0x5B
+    };
+
+    memcpy(&__osContPifRam[46], cicTest, 17);
+
+    __osSiRawStartDma(OS_WRITE, __osContPifRam);
+    osRecvMesg(&gSIEventMesgQueue, NULL, OS_MESG_BLOCK);
+    __osContLastCmd = 254;
+
+    __osSiRawStartDma(OS_READ, __osContPifRam);
+    osRecvMesg(&gSIEventMesgQueue, NULL, OS_MESG_BLOCK);
+    const u8 pifCheck = __osContPifRam[54];
+    __osSiRelAccess();
+
+    // check for the typo'd byte
+    return (pifCheck == 0xB0) ? EMU_PROJECT64_1_OR_2 : EMU_PROJECT64_3;
+}
+
+static u8 check_cache_emulation() {
+    // Disable interrupts to ensure that nothing evicts the variable from cache while we're using it.
+    u32 saved = __osDisableInt();
+    // Create a variable with an initial value of 1. This value will remain cached.
+    volatile u8 sCachedValue = 1;
+    // Overwrite the variable directly in RDRAM without going through cache.
+    // This should preserve its value of 1 in dcache if dcache is emulated correctly.
+    *(u8*)(K0_TO_K1(&sCachedValue)) = 0;
+    // Read the variable back from dcache, if it's still 1 then cache is emulated correctly.
+    // If it's zero, then dcache is not emulated correctly.
+    const u8 cacheEmulated = sCachedValue;
+    // Restore interrupts
+    __osRestoreInt(saved);
+    return cacheEmulated;
+}
+
+f32 round_double_to_float(f64 v) {
+    return v;
+}
+
+void detect_emulator() {
+    if ((u32)IO_READ(DPC_PIPEBUSY_REG) | (u32)IO_READ(DPC_TMEM_REG) | (u32)IO_READ(DPC_BUFBUSY_REG)) {
```

Comment:
Changed console check from
`IO_READ(DPC_CLOCK_REG) != 0`
to
`(u32)IO_READ(DPC_PIPEBUSY_REG) | (u32)IO_READ(DPC_TMEM_REG) | (u32)IO_READ(DPC_BUFBUSY_REG)`
to delay false detection of Ares as console

See https://discord.com/channels/879505781916053544/910581195891413002/1130950797614645329

Reply:
Is there any upside/downside to keeping in the DPC_CLOCK_REG check?

Reply:
You mean checking both? I guess it wouldn't cause any harm, but it also doesn't really help since if the emulator can actually emulate these correctly, I don't see how it also wouldn't implement `DPC_CLOCK_REG` correctly.

Reply:
Ah wait I've got it backwards, it's _console_ that can rarely falsely be detected as emulator and not the reverse. So probably no upside then.

Reply:
Wait no I had it right the first time lol, more checks means more chances for console to be properly detected, though I suppose there's three of them here already. But yeah adding back `DPC_CLOCK_REG` would break Ares detection so it's probably fine as is then.